File: src/client/acontext-py/src/acontext/uploads.py
"""
Utilities for working with file uploads.
"""

import io
from dataclasses import dataclass
from typing import BinaryIO


@dataclass(slots=True)
class FileUpload:
    """
    Represents a file payload for multipart requests.

    Accepts either a binary stream (any object exposing ``read``) or raw ``bytes``.
    """

    filename: str
    content: BinaryIO | bytes
    content_type: str | None = None

    def as_httpx(self) -> tuple[str, BinaryIO, str | None]:
        """
        Convert to the tuple format expected by ``httpx``.
        """
        if isinstance(self.content, (bytes, bytearray)):
            buffer = io.BytesIO(self.content)
            return self.filename, buffer, self.content_type or "application/octet-stream"
        return self.filename, self.content, self.content_type or "application/octet-stream"


def normalize_file_upload(
    upload: FileUpload | tuple[str, BinaryIO | bytes] | tuple[str, BinaryIO | bytes, str | None],
) -> FileUpload:
    if isinstance(upload, FileUpload):
        return upload
    if isinstance(upload, tuple):
        if len(upload) == 2:
            filename, content = upload
            return FileUpload(filename=filename, content=content)
        if len(upload) == 3:
            filename, content, content_type = upload
            return FileUpload(filename=filename, content=content, content_type=content_type)
    raise TypeError("Unsupported file upload payload")

File: src/client/acontext-ts/src/uploads.ts
/**
 * Utilities for working with file uploads.
 */

export class FileUpload {
  filename: string;
  content: Buffer | NodeJS.ReadableStream;
  contentType?: string | null;

  constructor(options: {
    filename: string;
    content: Buffer | NodeJS.ReadableStream;
    contentType?: string | null;
  }) {
    this.filename = options.filename;
    this.content = options.content;
    this.contentType = options.contentType ?? null;
  }

  /**
   * Convert to a format suitable for form-data.
   */
  asFormData(): {
    filename: string;
    content: Buffer | NodeJS.ReadableStream;
    contentType: string;
  } {
    return {
      filename: this.filename,
      content: this.content,
      contentType: this.contentType || 'application/octet-stream',
    };
  }
}

export function normalizeFileUpload(
  upload:
    | FileUpload
    | [string, Buffer | NodeJS.ReadableStream]
    | [string, Buffer | NodeJS.ReadableStream, string | null]
): FileUpload {
  if (upload instanceof FileUpload) {
    return upload;
  }
  if (Array.isArray(upload)) {
    if (upload.length === 2) {
      const [filename, content] = upload;
      return new FileUpload({ filename, content });
    }
    if (upload.length === 3) {
      const [filename, content, contentType] = upload;
      return new FileUpload({
        filename,
        content,
        contentType: contentType ?? null,
      });
    }
  }
  throw new TypeError('Unsupported file upload payload');
}

